pkg/secrets-store/utils.go

```diff
@@ -0,0 +1,727 @@
+		deleteSecretFn := func() (bool, error) {
+			item, err := getSecretProviderItemByName(ctx, secretProviderClass)
+			if err != nil {
+				log.Errorf("failed to get secret provider item, err: %v", err)
+				return false, nil
+			}
+			count := getStatusCount(item)
+			// only delete when no more pods are associated with it
+			if count == 0 {
+				///TODO: we assume all files are mounted from a single secretsproviderclass
+				/// a pod could have multiple volumes pointing to diff secretproviderclass objs
+				for _, s := range secretObjects {
+					secretObject, ok := s.(map[string]interface{})
+					if !ok {
+						continue
+					}
+					secretName, exists, err := unstructured.NestedString(secretObject, "secretName")
+					if err != nil {
+						return false, nil
+					}
+					if !exists {
+						continue
+					}
+					err = deleteK8sSecret(ctx, secretName, namespace)
+					if err != nil {
+						return false, nil
+					}
+				}
+			}
+			return true, nil
+		}
+
+		if err := wait.ExponentialBackoff(wait.Backoff{
+			Steps:    5,
+			Duration: 1 * time.Millisecond,
+			Factor:   1.0,
+			Jitter:   0.1,
+		}, deleteSecretFn); err != nil {
+			log.Error(err, "max retries for deleting secret reached")
+			return err
+		}
+	}
+	return nil
+}
+
+// createOrUpdateK8sSecret creates or updates a K8s secret with data from mounted files
+// If a secret with the same name already exists in the namespace of the pod, it's updated.
+func createOrUpdateK8sSecret(ctx context.Context, name string, namespace string, datamap map[string][]byte, secretType corev1.SecretType) error {
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	secretKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Type: secretType,
+	}
+
+	if err := c.Get(ctx, secretKey, secret); err != nil {
+		log.Error(err, "error from c.Get")
+		if errors.IsNotFound(err) {
+			secret.Data = datamap
+			if err := c.Create(ctx, secret); err != nil {
+				log.Error(err, "error while creating K8s secret")
+				return err
+			}
+		} else {
+			log.Error(err, "error while retrieving K8s secret")
+			return err
+		}
+	} else {
+		secret.Data = datamap
+		if err := c.Update(ctx, secret); err != nil {
+			log.Error(err, "error while updating K8s secret")
+			return err
+		}
+	}
+	log.Infof("created k8s secret: %s", name)
+	return nil
+}
+
+// deleteK8sSecret deletes a secret by name
+func deleteK8sSecret(ctx context.Context, name string, namespace string) error {
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	secretKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+	}
+
+	if err := c.Get(ctx, secretKey, secret); err != nil {
+		log.Error(err, "error while getting K8s secret")
+		return err
+	}
```

Comment:
If error from get is secret not found, can we return here?

Reply:
Right now this is returning for all errors. Can you clarify what you mean?

Reply:
nvm I got it